lib/client/src/client.rs

```diff
@@ -788,6 +898,142 @@ impl MangoClient {
+                && open_orders_data.native_coin_total == 0
+                && open_orders_data.native_pc_total == 0;
+
+            if is_closable {
+                serum3_closable_order_market_index = Some(p.market_index);
+                break;
+            }
+        }
+
+        let first_closable_slot =
+            serum3_closable_order_market_index.expect("couldn't find any serum3 slot available");
+
+        let ixs = PreparedInstructions::from_single(
+            self.serum3_close_open_orders_instruction(first_closable_slot),
+            self.context.compute_estimates.health_cu_per_serum,
+        );
+
+        let first_closable_market = account.serum3_orders(first_closable_slot)?;
+        let (tk1, tk2) = (
+            first_closable_market.base_token_index,
+            first_closable_market.quote_token_index,
+        );
+        account.token_position_mut(tk1)?.0.decrement_in_use();
+        account.token_position_mut(tk2)?.0.decrement_in_use();
+        account.deactivate_serum3_orders(first_closable_slot)?;
+
+        Ok(ixs)
+    }
+
+    async fn find_existing_or_try_to_replace_token_positions(
+        &self,
+        account: &mut MangoAccountValue,
+        token_indexes: Vec<TokenIndex>,
+    ) -> anyhow::Result<Option<PreparedInstructions>> {
+        let mut ixs = PreparedInstructions::new();
+
+        for token_index in token_indexes {
+            let result = self
+                .find_existing_or_try_to_replace_token_position(account, token_index)
+                .await?;
+            if let Some(ix) = result {
+                ixs.append(ix);
+            }
+        }
+        if ixs.is_empty() {
+            return Ok(None);
+        }
+
+        Ok(Some(ixs))
+    }
+
+    async fn find_existing_or_try_to_replace_token_position(
+        &self,
+        account: &mut MangoAccountValue,
+        token_index: TokenIndex,
+    ) -> anyhow::Result<Option<PreparedInstructions>> {
+        let token_position_missing = account
+            .ensure_token_position(token_index)
+            .is_anchor_error_with_code(MangoError::NoFreeTokenPositionIndex.error_code());
+
+        if !token_position_missing {
+            return Ok(None);
+        }
+
+        let ixs = self.deactivate_first_active_unused_token(account).await?;
+
+        Ok(Some(ixs))
+    }
+
+    async fn deactivate_first_active_unused_token(
+        &self,
+        account: &mut MangoAccountValue,
+    ) -> anyhow::Result<PreparedInstructions> {
+        let mut all_pos = Vec::new();
+        for x in account.all_token_positions() {
+            let bank = self.first_bank(x.token_index).await?;
+            let amount = x.native(&bank);
+            let name = bank.name().to_string();
+            all_pos.push((x.token_index, x.in_use_count, amount, name));
+        }
+
+        let closable_tokens = account
+            .all_token_positions()
+            .enumerate()
+            .filter(|p| p.1.is_active() && !p.1.is_in_use());
```

Comment:
just fyi, another way to write this would be `|(_, p)| p.is_active() && !p.is_in_use()` which I weakly prefer